web-common/src/features/sources/modal/submitAddDataForm.ts

```diff
@@ -123,14 +115,85 @@ export async function submitAddDataForm(
     createOnly: false,
   });
 
-  // Update the `.env` file
+  // Check for an existing `.env` file
+  // Store the original `.env` blob so we can restore it in case of errors
+  let originalEnvBlob: string | undefined;
+  try {
+    const envFile = await queryClient.fetchQuery({
+      queryKey: getRuntimeServiceGetFileQueryKey(instanceId, { path: ".env" }),
+      queryFn: () => runtimeServiceGetFile(instanceId, { path: ".env" }),
+    });
+    originalEnvBlob = envFile.blob;
+  } catch (error) {
+    const fileNotFound =
+      error?.response?.data?.message?.includes("no such file");
+    if (fileNotFound) {
+      // Do nothing. We'll create the `.env` file below.
+    } else {
+      // We have a problem. Throw the error.
+      throw error;
+    }
+  }
+
+  // Create or update the `.env` file
+  const newEnvBlob = await updateDotEnvWithSecrets(
+    queryClient,
+    connector,
+    formValues,
+    "connector",
+  );
   await runtimeServicePutFile(instanceId, {
     path: ".env",
-    blob: await updateDotEnvWithSecrets(queryClient, connector, formValues),
+    blob: newEnvBlob,
     create: true,
     createOnly: false,
   });
 
+  /**
+   * Test the new OLAP connector:
+   * 1. Ensure the file has reconciled and has no errors
+   * 2. Test the connection to the OLAP database
+   */
+
+  // Wait until the project parser has reconciled the new connector file
+  await waitForProjectParserVersion(
+    instanceId,
+    projectParserStartingVersion + 1,
```

Comment:
Saving `projectParserStartingVersion` just to wait for next version seems weird in the caller function.

Could we add a version of `runtimeServicePutFile` that hides the details from the caller?